lib/SILOptimizer/Analysis/RCIdentityAnalysis.cpp

```diff
@@ -58,14 +58,31 @@ static bool isRCIdentityPreservingCast(ValueKind Kind) {
   }
 }
 
+/// Returns a null SILValue if V has a trivial type, otherwise returns V.
+///
+/// RCIdentityAnalysis must not look through casts which cast from a trivial
+/// type, like a metatype, to something which is retainable, like an AnyObject.
+/// On some platforms such casts dynamically allocate a ref-counted box for the
+/// metatype.
+/// Now, if the RCRoot of such an AnyObject would be a trivial type, ARC
```

Comment:
A few concerns and then a proposed solution:

1. The prefix 'no' makes this sound like a boolean predicate which is misleading. We really should find a different name. That suggests to me that maybe this is too low level of granularity (see below).

2. The comment here I think can be made clearer by saying that because the cast dynamically allocates a uniquely identified ref-counted box, naturally that is the place where a new rc-identity begins. That doesn't mean that we shouldn't remove retain, release on these objects, but that we should consider the cast to be introducing a new rc identical object. I am not sure if we should consider it to be a +1 value but I imagine it should be given that we are allocating a new boxed object.

With those two in hand, after some thought I wonder if maybe the more appropriate thing to do would be to change isRCIdentityPreservingCast to take a SILValue so that we have access to the underlying type info. Then we can add the comment I mentioned above perhaps. Maybe make it more about the result of the cast and have a helper routine that grabs a source... IDK.

Reply:
On another note, I also just did a git grep and found in InstructionUtils.cpp we have another implementation of this. We probably want to centralize or fix this.

Reply:
```
lib/SIL/Utils/InstructionUtils.cpp:static bool isRCIdentityPreservingCast(ValueKind Kind) {
lib/SIL/Utils/InstructionUtils.cpp:    if (isRCIdentityPreservingCast(K)
lib/SIL/Utils/InstructionUtils.cpp:    if (isRCIdentityPreservingCast(k)
lib/SILOptimizer/Analysis/RCIdentityAnalysis.cpp:static bool isRCIdentityPreservingCast(ValueKind Kind) {
lib/SILOptimizer/Analysis/RCIdentityAnalysis.cpp:  if (isRCIdentityPreservingCast(V->getKind()))
```